\documentclass[12pt]{article}
\usepackage{paper,math}
\usepackage[margin=1in]{geometry}
\addbibresource{references.bib}
\usepackage[linesnumbered,ruled,vlined]{algorithm2e}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{float}

\title{Gene Set Enrichment Analysis \\Assignment 5\\ Genomorientierte Bioinfromatik}
\author{Malte Weyrich}
\date{\today}
% Conditionally display thoughts (hide by switching to `\boolfalse`)
% \booltrue{INCLUDECOMMENTS}
\newcommand{\malte}[1]{\coauthorComment[Malte]{#1}}

\begin{document}

% Title Page -------------------------------------------------------------------
\maketitle

\tableofcontents
\begin{abstract}
When performing \textit{Enrichment Analysis} in bioinformatics, there are several different tools and methods available to determine whether certain gene sets are statistically enriched. 
These tools usually approach this problem in different ways, using different statistical tests and assumptions.
The two main methods are \textit{Gene Set Over-representation Analysis (ORA)} and \textit{Gene Set Enrichment Analysis (GSEA)}.
This report compares a custom Java implementation of both methods for a given list of differentially expressed genes and a standard of truth gene set,
with two already existing tools, called \textit{gProfiler (perfoms ORA)} and \textit{fgsea ( performs GSEA)}.
The Java implementation is also evaluated in terms of runtime and the properties of the \textit{Directed Acyclic Graph (DAG)} used for the analysis.


\end{abstract}

\newpage


% Paper ------------------------------------------------------------------------

% ------------------------------------------------------------------------------
\section{Introduction}\label{sec:Introduction}
Due to the vast amount of different genes and gene products, a vocabulary for describing
groups of genes and their logical relationships called \textit{\textbf{Gene Ontology
(GO)}}, was invented by the \textit{\textbf{Gene Ontology Consortium (GOC)}} (\cite{Ashburner2000}).
In \textit{Gene Ontology}, genes are sorted into different \textit{gene sets} (referred
to as \textit{GO Terms}), with each gene set having its own ID.
These gene sets are annotated with a function and a process
in which the genes of the set are involved. A gene set can be a component of several
other gene sets through an \textit{is\_a relationship}, which leads to a hierarchical,
tree-like structure. In this case, the structure is a \textit{\textbf{Directed Acyclic
Graph (DAG)}}. The biological functions and processes are sorted into three main categories:
\textit{Biological Process (BP)}, \textit{Molecular Function (MF)} and \textit{Cellular
Component (CC)}, each forming its own \textit{DAG}. The root of each \textit{DAG}
comprises all genes of the \textit{DAG} and is labeled with the one of the \textit{GO
Terms}. The amount of genes per gene set decreases the further its location is
relative to the root. The same goes for the annotations of gene sets, which get more exact
and detailed the further they are away from the root. 

In bioinformatics, this vocabulary is crucial for determining the effects of differentially
expressed genes in RNA-seq and other experiments. Knowing which genes are significantly
up or down-regulated is not sufficient enough for understanding the biological processes
that are affected by the changes in gene expression. This is where \textit{\textbf{Gene
Set Enrichment Analysis (GSEA)}} and \textit{\textbf{Gene Over-representation Analysis
(ORA)}} come into play. Although these two methods differ in their methodology, the
main goal is to determine whether certain gene sets are statistically enriched, providing insights into affected biological pathways.

The analysis conducted in this report tries to compare the results of both methods
for a given list of differentially expressed genes and standard of truth gene sets.
Additionally, our results are compared to two already existing tools for \textit{ORA}, called
\textit{\textbf{gProfiler}} (\cite{gprofiler}) and \textit{\textbf{fgsea}} (\cite{fgsea})
for \textit{GSEA}.

\section{Materials \& Methods}
\subsection{Enrichment Methods}\label{sec:Enrichment-Methods}
\subsubsection{Gene Set Over-representation Analysis}\label{sec:Gene-Set-Over-representation-Analysis}
In \textit{ORA}, the first step is to define a gene list $L$ of \textbf{possible interesting genes} (genes having $padj \le 0.05$ or $abs(lfc) \ge 1.5$).
Then, for each gene set $S$, an \textit{enrichment p-value} is calculated, representing the 
probability of observing at least as many genes from $L$ in $S$, assuming a random selection from the background gene set (\cite{ORA}).
This is commonly simulated using either a \textit{Hypergeometric Test} or the \textit{Fisher's Exact Test}.
In our case, we used \textit{Fisher's Exact Test} with the following parameters:
\[
    P(X \ge k) = 1 - \frac{\binom{K}{k} \cdot \binom{N-K}{n-k}}{\binom{N}{n}} 
\]
, where 
\begin{itemize}
    \item $N := |G \cap DE|$ 
    \item $n := |g \cap DE|$ 
    \item $K := |G \cap sDE|$ 
    \item $k := |g \cap sDE|$ 
\end{itemize}
, with 
\begin{itemize}
    \item $G := $ All gene symbols of current \texttt{DAG}
    \item $g := $ Gene symbols of current \texttt{GOEntry}
    \item $DE := $ All observed differentially expressed genes 
    \item $sDE :=$ Subset of $sDE \subseteq DE$ considered significant
\end{itemize}

After calculating all \textit{enrichment p-values}, they are corrected for multiple testing
using the \textit{Benjamini-Hochberg} method.

One problem in \textit{ORA} is that we have to define the threshold and choose, whether we only want to
analyze up-regulated or down-regulated genes, or both, which is not the case in \textit{GSEA}.

\subsubsection{Gene Set Enrichment Analysis}\label{sec:Gene-Set-Enrichment-Analysis}
\textit{GSEA} use a different approach to determine the significance of gene sets.
It consists of three main steps:
\begin{enumerate}
    \item Calculate the \textit{Enrichment Score (ES)} of a gene set
    \item Evaluate the significance of the calculated Enrichment Score (through \textit{empirical p-value})
    \item Correct the p-values for multiple testing (using \textit{Benjamini-Hochberg})
\end{enumerate}

\hspace{1mm}\\
In order to calculate the \textit{ES} for a given gene set $S$,
the observed differentially expressed genes of an experiment are ranked
based on their \textit{logFoldChange}, resulting in a \textit{ranked list} $L$.
Important here is that all observed genes are used, whereas in 
\ref{sec:Gene-Set-Over-representation-Analysis} only the significantly differentially 
expressed genes are used. 
The \textit{ES} is then calculated by walking down the ranked list $L$ and
calculating a \textit{running sum statistic} which increases when a gene
is part of the gene set $S$ and decreases when it is not.
The maximum value that is encountered during the walk is the \textit{ES} of the gene set $S$.
The \textit{running sum statistic} utilizes a \textit{Kolmogorov-Smirnov} like statistic 
that weights genes according to their \textit{logFoldChange} magnitude (\cite{GSEA}).

The provided \textit{JAR} imports the \texttt{KolmogorovSmirnovTest Class} from \\
\textit{org.apache.commons.math3.stat.inference} in order to calculate the
\textit{Kolmogorov-Smirnov} statistic and p-value.
We do not calculate the \textit{empirical p-value} for the \textit{ES} as it was 
not required in the assignment, but usually, it is more than required to
run a permutation test to determine the true significance of the \textit{ES}
in order to avoid too many false positives.

\newpage
\subsection{Java Programm}\label{sec:Java-Programm}
The provided \textit{JAR} has the following input specification:
\begin{verbatim*}
java -jar go_enrichment.jar
  -obo                   Path to obo file.
  -root                  One of three Ontology Terms:
                         ["molecular_function",
                         "biological_process",
                         "cellular_component"].
  -mapping               Path to mapping file (SYMBOL -> GO_ID).
  -mappingtype           Format of the mapping-File (go|ensembl). 
                         Has to agree with "-mapping" option.
  [-overlapout]          Information about DAG  entries  with shared mapped
                         genes is written into this file.
  -enrich                Path to enrichment analysis file.
  -o                     Path to output file.
  -minsize               Min amount of genes per GO entry.
  -maxsize               Max amount of genes per GO entry.
\end{verbatim*}

The specific file formats are described in \textit{Assignment 5}.
\newpage
\subsubsection{Logic}\label{sec:Logic}
\textbf{(A) Parsing Input Files:}

The provided \textit{JAR} reads in the necessary input files and stores them for later use.
Parsing the \textit{obo} file is the first step since it contains the structure
of the \texttt{DAG}, which itself is an object. The parser only considers entries that map
the parameter \textit{"-root"} and ignores entries that are marked as obsolete.
It stores entries in a \texttt{HashMap<String, GOEntry> nodeMap} to keep 
track of already-seen gene sets. For each entry, we check if
its GO ID is contained in our \texttt{nodeMap}. If not, we append
a new \texttt{GOEntry} object to \texttt{nodeMap} and add its parents 
to the newly created object. Otherwise, we take the already existing \texttt{GOEntry} 
object inside the \texttt{nodeMap} and update its parents. 
Of course, we also have to look up parent GO IDs inside \texttt{nodeMap}
before creating a new \texttt{GOEntry} object. This way we can ensure that 
the \texttt{DAG} is correctly built up. The \texttt{root} of the \texttt{DAG} is
the obo entry which has the same name as the parameter \textit{"-root"}.

Based on the provided \textit{"-mappingtype"}, the \textit{JAR} has two different 
methods for parsing the file specified in \textit{"-mapping"}. 
This part populates the created \texttt{DAG} object with the gene symbols.
The symbols are stored as a \texttt{HashSet<String> geneSymbols} in each \texttt{GOEntry} object.
Lastly, the \textit{JAR} reads in the enrichment analysis file and stores the results
in a \texttt{HashMap<String, Gene> enrichedGeneMap}. The first lines of our enrichment file
also contains "Standard of Truth" (SoT) GO ids, which are gene sets that are known to be
enriched for the given list of differentially expressed genes. These SoT GO ids are 
used to mark the corresponding \texttt{GOEntry} objects in the \texttt{DAG} and are
additionally stored in a \texttt{HashSet<GOEntry> trueGoEntries}. We will use this
for later comparisons.\\
\hspace{1mm}
\hspace{1mm}
\textbf{(B) Preparing DAG for Enrichment Analysis:}

At this point of the program, we have a fully constructed \texttt{DAG} object.
As described in \ref{sec:Introduction}, a parent \texttt{GOEntry} object
inherits all genes from its children.
Calling the method \texttt{inheritGeneSymbols()} on the \texttt{root} of the \texttt{DAG}
leads to a recursive update of the gene symbols in each \texttt{GOEntry} object
from the bottom to the top of the \texttt{DAG}. After this step, the \texttt{root}
contains all gene symbols of the \texttt{DAG}. The same logic is used to 
create a \texttt{HashSet<String> reachableGOIDs} in each \texttt{GOEntry} object,
which will be useful for later calculations. The result file should also contain
a field called \textit{"shortest\_path\_to\_a\_true"}. This field is supposed to
describe the shortest path from a \texttt{GOEntry} $A$ to the closest true
\texttt{GOEntry} $B$. 
These paths can be pre-computed, avoiding expensive breadth-first searches which
might compute the same path multiple times.
For this purpose, each \texttt{GOEntry} object has a 
field \texttt{HashMap<GOEntry, LinkerClass> HighwayMap} which stores the
true \texttt{GOEntries} \textbf{as keys} and a \texttt{LinkerClass} object as value.
The \texttt{LinkerClass} object contains the \texttt{GOEntry} object
\textbf{which has to be traversed} to reach the true \texttt{GOEntry} and
a value $k$ which describes the remaining distance to the true \texttt{GOEntry}.

\begin{figure}[htpb]
    \centering
    \includegraphics[width=0.99\textwidth]{./figures/Highway.png}
    \caption{A simple example of how the \texttt{HighwayMap} uses the true \texttt{GOEntries} as key and stores a \texttt{LinkerClass} object as value,
    serving as a highway to the true \texttt{GOEntry}. In this case there's only one \texttt{GOEntry} marked as\textit{true}, that's why the
    \texttt{HighwayMap} only contains one element per \texttt{GOEntry}. The figure was created using \cite{biorender}.}
    \label{fig:-figures-Highway-png}
\end{figure}


We start populating the \texttt{HighwayMap} with the true \texttt{GOEntries}
by fist calling the method \texttt{signalShortestPathUp(0, GOEntry trueGO)} (Algorithm \ref{alg:signalShortestPathUp}) on each true \texttt{GOEntry}.
This method is called recursively on each parent of the true \texttt{GOEntry} and
propagates the shortest path to the true \texttt{GOEntry} upwards in the \texttt{DAG}.
This alone is insufficient since we also have to propagate the shortest path
downwards in the \texttt{DAG}. 
For this purpose, we call the method \texttt{propagateShortestPaths(trueGo)} (Algorithm \ref{alg:propagateShortestPaths}) on the \texttt{\textbf{root}} 
of the \texttt{DAG}.
While the method \texttt{propagateShortestPaths()} ensures that the shortest paths propagate 
down from the \texttt{root}, it does not explicitly finalize the correct shortest 
paths for every node. This is done by calling the method \texttt{signalShortestPathDown(0, trueGo)} (Algorithm \ref{alg:signalShortestPathDown}) on all true \texttt{GOEntries}.
It makes sure that once the shortest paths are propagated down from the \texttt{root}, 
every node (especially those closer to \texttt{trueGo}) has the correct, 
finalized shortest path to the trueGo.

\begin{algorithm}[!htbp]
\caption{signalShortestPathUp(k, trueGo)}\label{alg:signalShortestPathUp}
\KwIn{$k$, trueGo}
\KwOut{Updated shortest path values for all parents}
\ForEach{parent in parents}{
    \If{parent.highwayMap does not contain trueGo}{
        link = new LinkerClass(this, k + 1) \;
        parent.highwayMap.put(trueGo, link) \;
    }
    \ElseIf{parent.highwayMap.get(trueGo).getK() > k + 1}{
        parent.highwayMap.get(trueGo).setGo(this) \;
        parent.highwayMap.get(trueGo).setK(k + 1) \;
    }
    parent.signalShortestPathUp(k + 1, trueGo) \;
}
\end{algorithm}


\begin{algorithm}[!htbp]
\caption{propagateShortestPaths(trueGo)}\label{alg:propagateShortestPaths}
\KwIn{trueGo}
\KwOut{Updated shortest path values for all descendants}
currentPath $\gets$ highwayMap.get(trueGo) \;
\If{currentPath is not null}{
    nextNode $\gets$ currentPath.getGo() \;
    \If{nextNode.highwayMap contains trueGo}{
        pathThroughLinker $\gets$ nextNode.highwayMap.get(trueGo).getK() + 1 \;
        \If{pathThroughLinker < currentPath.getK()}{
            currentPath.setK(pathThroughLinker) \;
        }
    }
}
currentK $\gets$ currentPath is not null ? currentPath.getK() : \texttt{Integer.MAX\_VALUE} \;
\ForEach{child in children}{
    childPath $\gets$ child.highwayMap.get(trueGo) \;
    childK $\gets$ childPath is not null ? childPath.getK() : \texttt{Integer.MAX\_VALUE} \;
    \If{currentK is not \texttt{Integer.MAX\_VALUE} and currentK + 1 < childK}{
        \If{childPath is null}{
            childPath $\gets$ new LinkerClass(this, currentK + 1) \;
            child.highwayMap.put(trueGo, childPath) \;
        }
        \Else{
            childPath.setGo(this) \;
            childPath.setK(currentK + 1) \;
        }
    }
    child.propagateShortestPaths(trueGo) \;
}
\end{algorithm}


\begin{algorithm}[!htbp]
\caption{signalShortestPathDown(k, trueGo)}\label{alg:signalShortestPathDown}
\KwIn{$k$, trueGo}
\KwOut{Updated shortest path values for all descendants}
k $\gets$ k + 1 \;
\ForEach{child in children}{
    \If{child.highwayMap does not contain trueGo}{
        link $\gets$ new LinkerClass(this, k) \;
        child.highwayMap.put(trueGo, link) \;
    }
    \ElseIf{child.highwayMap.get(trueGo).getK() > k}{
        child.highwayMap.get(trueGo).setGo(this) \;
        child.highwayMap.get(trueGo).setK(k) \;
    }
    child.signalShortestPathDown(k, trueGo) \;
}
\end{algorithm}

\newpage
\hspace{1mm}\\
\textbf{(C) Enrichment Analysis:}

The actual enrichment analysis is conducted in the method \texttt{analyzeParallel()} of the
class \texttt{EnrichmentAnalysis}. This method iterates over all \texttt{GOEntry} objects
and checks whether their amount of genes is within the specified range of \textit{"-minsize"}
and \textit{"-maxsize"}. 
An object called \texttt{AnalysisEntry} is created for each \texttt{GOEntry} object which
fulfills the requirements.
It then calculates the overlap of the gene symbols of the current \texttt{GOEntry} with 
the gene symbols of the differentially expressed genes and stores the result 
in the \textit{"size"} field of the \texttt{AnalysisEntry} object ($\equiv n$).
The attribute \textit{"noverlap"} is the amount of significantly differentially expressed genes
which also occur in the current \texttt{GOEntry} ($\equiv k$). 
The universe size is the intersection of all gene symbols of the differentially expressed genes
and all gene symbols of the \texttt{DAG} ($\equiv N$). Lastly, $K$ is the amount of significantly
differentially expressed genes intersected with all gene symbols of the \texttt{DAG}.
For the Kolmogorow-Smirnow-Test we define the following values:
\begin{itemize}
    \item[\textbf{I.}] \textit{In-Set Distribution}: Genes present in the current \textit{GOEntry} and differentially expressed genes.
    \item[\textbf{II.}] \textit{Background Distribution}: Genes not in the \textit{In-Set} but part of the background gene pool.
\end{itemize}
For both of these distributions, we store the \textit{logFoldChange} values of the differentially
expressed genes into two arrays.
With these values, the statistical tests described in \ref{sec:Enrichment-Methods} are 
conducted to determine the significance of the enrichment of the current \texttt{GOEntry}. 
As a result, we obtain the following values and store them in the \texttt{AnalysisEntry} object:
\begin{itemize}
    \item \textit{hgPval}: Hypergeometric p-value.
    \item \textit{fejPval}: Jackknife Fisher's Exact Test p-value.
    \item \textit{ksPval}: Kolmogorov-Smirnov p-value.
    \item \textit{ksStat}: Kolmogorov-Smirnov statistic ($\equiv$ \textit{Enrichment Score}).
\end{itemize}
\hspace{1mm}\\
Now we calculate the shortest path to the closest true \texttt{GOEntry} for each \texttt{AnalysisEntry}
by utilizing the pre-computed \texttt{HighwayMap} of the current \texttt{GOEntry}.
We simply choose the \texttt{GOEntry} object which points to the \texttt{LinkerClass}
with the smallest $k$ value in the \texttt{HighwayMap} as our \texttt{target GOEntry}. 
For each visited \texttt{GOEntry} object on the path, we check
if the \texttt{HighwayMap} still contains the \texttt{target GOEntry} as key.
If so, we follow its \texttt{LinkerClass} object until we reach the \texttt{target}.
We know when we hit the \textit{least common ancestor (LCA)} of the current \texttt{GOEntry}
and the closest true \texttt{GOEntry},
by checking if the \texttt{reachableGOIDs} object contains our \texttt{target GOEntry}
for each visited \texttt{GOEntry} object on the current path.

After all \texttt{AnalysisEntry} objects have been processed, 
all p-values are corrected for multiple testing using the \textit{Benjamini-Hochberg}
method. The corrected p-values are stored in the \texttt{AnalysisEntry} objects
, and each object is written to the output file.
The above procedure is conducted in parallel in order to speed up the process.
\hspace{1mm}\\
\textbf{(D) GO Features:}

The optional parameter \textit{"-overlapout"} specifies an additional 
output file that stores further features of the underlying \texttt{DAG}.
For this, each unique pair of \texttt{GOEntries} $(A, B)$ fulfilling the
\textit{"-minsize"} and \textit{"-maxsize"} parameter are 
intersected by their \textit{gene symbols}. If there is an overlap,
we calculate the shortest path between $A$ and $B$.
In this case, we use a \textit{Breadth-First Search} algorithm
to explore all ancestors of $A$ and $B$ and return the shortest path.
Additionally, we write a \texttt{boolean} value into the output file, indicating whether $A$ can be directly reached from $B$ or vice versa.
Lastly, we store the max percentage of shared genes between $A$ and $B$:
\[
    \textit{max percentage} = \frac{|A \cap B|}{\textit{max}\left\{|A|, |B|\right\}} \times 100
.\]

\section{Results}
\subsection{Runtime}\label{sec:Runtime}
The runtime of the  \textit{JAR} was timed using a \texttt{Logger Class}, which 
timed certain analysis steps. The two main analysis methods are implemented as 
single-threaded and multi-threaded and will be compared in this section. We used the provided
\textit{obo} file, \textit{mapping} file and \textit{enrichment} file for the analysis and ran the \textit{JAR} with 
\textit{"-minsize" 50} and \textit{"-maxsize" 500}. 

Figure \ref{fig:-plots-times-png} shows the total runtime of the \textit{JAR} and
also splits it into several sub-tasks like parsing the input files, preparing the
\texttt{DAG} for the enrichment analysis itself. Running
the \textit{JAR} with \textit{"-mappingtype ensembl"} seems to be faster compared
to the \textit{"go"} mapping. This holds for almost every category, except for "optimizing DAG",
which takes 0.01s longer, but this is most likely a hardware inconsistency.
The \textit{"ensembl"} mapping file has fewer genes mapped to \texttt{GOEntries} than the 
\textit{"go"} mapping file, and because the runtime of every step is highly dependent
on the number of genes inside a \textit{gene set}, having fewer genes in total also leads to a faster
runtime.
\begin{figure}[htpb]
    \centering
    \includegraphics[width=0.8\textwidth]{./plots/times.png}
    \caption{Comparison the time duration of key setps of the program based on\\ \textit{"-mappingtype"} and implementation variant.
    Note that only the last two steps, "Enrichment Analysis" and "GO Features", were parallellized.}
    \label{fig:-plots-times-png}
\end{figure}

The two analysis steps "Enrichment Analysis" (\textbf{step (C)}) and "GO Features" (\textbf{step (D)})
seem to scale ruffly proportional between \textit{"-mappingtype"} and implementation type.
An interesting aspect is that parallelizing step (D) seems to save less time compared 
to the parallelization of step (C). This is because of the large amount of set operations 
required during step (C). For the parameters used in the evaluation, 
a total of $\frac{n(n - 1)}{2}$ many unique pairs have to be compared by intersecting their
gene symbols. 
In this case $n$ is the number of \texttt{GOEntries} of the \texttt{DAG} with a gene set 
size $|g|$ of $50 \le |g| \le 500$. This results in 3 378 224 pairs with the \textit{"go
mappingtype"} and 2 483 776 pairs with the \textit{"ensembl mappingtype"},
whereas in step (C), only 1838 ("go") or 1576 ("ensembl") many \texttt{GOEntries} 
are analyzed.

\newpage
\subsection{DAG Properties}\label{sec:DAG-Properties}

\begin{table}[htpb]
    \centering
    \setlength{\tabcolsep}{4pt} % Adjust column padding
    \caption{DAG Properties summary for "-minsize 50" and "-maxsize 500". The table shows the number of genes, gene sets, leaves, the length of the shortest $S$ and longest $L$ path to the \texttt{root} in the DAG and the amount of analyzed \texttt{GOEntries} which 
    agree with the "-minsize" and "-maxsize" parameters.} \label{tab:dag_properties}
    \small % Reduce font size
\begin{tabular}{|l|r|r|r|r|l|l|} \hline
     \textbf{Mapping Type} & \#Genes & \#\texttt{GOEntries} & \#GOs analyzed & \#Leafs & $S$ & $L$ \\ \hline
     \textbf{GO} & 17026 & 29385 & 1838 & 14991 & GO:0031629 $\to$ 2 & GO:1905741 $\to$ 17 \\ \hline
     \textbf{ENSEMBL} & 15496 & 29385 & 1576 & 14991 & GO:0031629 $\to$ 2 & GO:1905741 $\to$ 17 \\ \hline
\end{tabular}
\end{table}

Table \ref{tab:dag_properties} shows the properties of the \texttt{DAG} for both mapping
types. The \texttt{DAG} itself contains 29385 gene sets, with 14991 leaves. The shortest
path to the \texttt{root} is 2 edges long, while the longest path is 17 edges long.
These properties are the same for both mapping types since the \texttt{DAG} structure remains the
same for both "go" and "ensembl". The only difference is the gene symbols which are
stored in the \texttt{GOEntries} (15496 vs. 17026) and the amount of \texttt{GOEntries} 
analyzed.
Figure \ref{fig:-plots-goSizes-png} additionally shows the distribution of gene set sizes in the \texttt{DAG} for both mapping types,
as well as the difference in gene set sizes between parent and child nodes.
It seems that the difference in gene set size per traversed edge is slightly more pronounced for the \texttt{GO} mapping type compared to the \texttt{"ensembl"} mapping type,
while the overall distribution of gene set sizes is similar for both mapping types.

\begin{figure}[!htbp]
    \centering
    \includegraphics[width=0.8\textwidth]{./plots/goSizes.png}
    \caption{Comparison of gene set sizes in the DAG as well as the difference in gene set sizes between parent and child nodes, per mapping type.}
    \label{fig:-plots-goSizes-png}
\end{figure}

\newpage
\subsection{Enrichment Results}
Before we analyzed the results of the enrichment analysis, we filtered
the 19 \textit{SoT} gene sets by their size, leaving us with 13 \textit{SoT's}
which could be discovered by our \textit{JAR}.


\begin{figure}[htpb]
    \centering
    \begin{minipage}{0.45\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/BHBoxplotFDRFEJ.png}
        \caption{Distribution of BH adjusted (Jackknife) Fisher Exact Test p-values. }
        \label{fig:boxplot-fdr-fej}
    \end{minipage}
    \hfill
    \begin{minipage}{0.45\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/BHBoxplotFDRKS.png}
        \caption{Distribution of BH adjusted Kolmogorov-Smirnov p-values.}
        \label{fig:boxplot-fdr-ks}
    \end{minipage}
\end{figure}

Figure \ref{fig:boxplot-fdr-fej} and \ref{fig:boxplot-fdr-ks} show the distribution of the
adjusted p-values, where the top 5 most significant \texttt{GOEntries} are labeled for each 
"-mappingtype". SoT gene sets are colored in blue.
Right away, a major discrepancy is that there are way more significant results 
for the \textit{ORA} (Figure \ref{fig:boxplot-fdr-fej}) approach compared to 
the \textit{GSEA} approach in Figure \ref{fig:boxplot-fdr-ks}.
The $\alpha$ threshold of 0.05 is leveling with the $Q3$ of the boxplot, meaning
that the majority of the significant results are below the $\alpha$ threshold,
even after multiple testing correction.

\begin{figure}[htpb]
    \centering
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/ensScatt.png}
        \caption{Ensembl Mapping Scatter Plot: significance vs gene set size.}
        \label{fig:ens-scatt}
    \end{minipage}
    \hfill
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/goScatt.png}
        \caption{GO Mapping Scatter Plot: significance vs gene set size.}
        \label{fig:go-scatt}
    \end{minipage}
\end{figure}

We can also see how smaller gene sets are more often considered significant
for both \textit{"-mappingtype" options}, as shown in Figure \ref{fig:ens-scatt} and \ref{fig:go-scatt}.
The \textit{"go"} mapping type has a higher amount of significant gene sets compared to the \textit{"ensembl"} mapping type
because the overall amount of genes in the \texttt{DAG} is higher (see Table \ref{tab:dag_properties}).

\begin{figure}[htpb]
    \centering
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/gseaOverlap.png}
        \label{fig:gsea-overlap}
    \end{minipage}
    \hfill
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/oraOverlap.png}
        \label{fig:ora-overlap}
    \end{minipage}
    \caption{Comparison of the results of the \textit{GSEA} and \textit{ORA} approach based on 
    the \textit{"-mappingtype"}. Only \texttt{GOEntries} with a padj below 0.05 are considered.
    In this case, ORA uses the adjusted Jackknife Fisher Exact Test (fej-pval), and GSEA uses the adjusted Kolmogorov-Smirnov p-value.}
\end{figure}

Again, we can see how the \textit{GSEA} approach seems to be more conservative compared to the \textit{ORA} approach,
since there are less significant results for the \textit{GSEA} approach.
Both approaches identify almost all SoT gene sets, which is a good sign.
For \textit{ORA} and \textit{GSEA}, two of the SoT gene sets are not found, which are \textit{"GO:0098754
"} and \textit{"GO:2000242"}. These gene sets are non-existent in the \textit{"ensembl mapppingtype"} 
and are therefore not discovered.
\begin{verbatim*}
→ grep -e "GO:2000242|GO:0098754" goa_human_ensembl.tsv | wc -l
0                                                              
\end{verbatim*}


\begin{figure}[htpb]
    \centering
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/ensOverlap.png}
    \end{minipage}
    \hfill
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/goOverlap.png}
    \end{minipage}

    \caption{Comparison of overlap of significant \texttt{GOEntries} of same \textit{"-mappingtype"} between ORA and GSEA.}
    \label{fig:overlap}
\end{figure}

Figure \ref{fig:overlap} shows how, in both "-mappingtype" cases, almost all significant
\texttt{GOEntries} discovered by \textit{GSEA}, are also discovered by \textit{ORA}.
For the \textit{"ensembl"} mapping, \textit{GSEA} finds 6 entries not covered by \textit{ORA}.
For the "go" mapping, this number is reduced to 3.
This means that in our case, \textit{GSEA} and \textit{ORA} are mostly aligned with each other,
with the exception of \textit{GSEA} having less \textit{FPs} than \textit{ORA}.
The actual amount of significant results of \textit{GSEA} is likely even less than what is shown
in Figure \ref{fig:overlap}, since, as mentioned in Section \ref{sec:Gene-Set-Enrichment-Analysis},
the \textit{enrichment p-vals} are not properly evaluated through \textit{phenotype permutation} (\cite{GSEA}).

\subsection{Comparison with Online Tool gProfiler (ORA)}\label{sec:Comparison-of-ORA}
For the comparison with \textit{gProfiler}, we uploaded the list of all \textit{significant} genes
of the \textit{enrichment file} to the web application. \textit{gProfiler} performs
\textit{ORA}, using known gene sets and pathways from several different databases.
We filtered for gene sets complying with out \textit{"-minsize 50"} and \textit{"-maxsize 500"}
parameters (see Figure \ref{fig:-screenshots-gProfiler-png}) and only looked at 
the \textit{BP Ontology}. 
\begin{figure}[H]
    \centering
    \includegraphics[width=0.8\textwidth]{./screenshots/gProfiler.png}
    \caption{Filtering \textit{gProfiler} results to match \textit{"-minsize"} and \textit{"-maxsize"} parameters.}
    \label{fig:-screenshots-gProfiler-png}
\end{figure}
\begin{figure}[H]
    \centering
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/go_mappingCompgProfiler.png}
    \end{minipage}
    \hfill
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/ens_mappingCompgProfiler.png}
    \end{minipage}
    \caption{Comparing significantly enriched gene sets discovered by gProfiler and our JAR, based on "-mappingtype".} 
    \label{fig:comp-gProfiler}
\end{figure}
We can see how in Figure \ref{fig:comp-gProfiler}, \textit{gProfiler} identifies even more 
significant gene sets than our \textit{JAR}.
At the same time, the external tool seems to struggle with 4 \texttt{GOEntries} in both cases
which could be due to a similar issue where GO IDs of the \textit{SoT} list are not
present in the \texttt{GOEntries} \textit{gProfiler} uses for its analysis.
\newpage

\subsection{Comparison with R library fgsea (GSEA)}\label{sec:Comparison-of-GSEA}
We ran the library \textit{fgsea} on our data in order to compare the \textit{GSEA} analysis
of our \textit{JAR} with a state-of-the-art implementation.
The results are shown in Figure \ref{fig:go-mapping-fgsea} and show that, as expected,  \textit{fgsea}
finds less significant gene sets compared to our \textit{JAR}. This is due to the fact that
\textit{fgsea} computes the empirical p-value for the \textit{ES} (\cite{fgsea}) and therefore avoids too many
FPs. 
As gene sets we used the library \textit{msigdb} (\cite{msigdb2024}) with the following specifications:
\begin{verbatim*}

# read provided enrichment file
genes <- fread("genes.tsv")
#      id      fc signif
#           <char>   <num> <lgcl>
# 1: DNAJC25-GNG10 -1.3420  FALSE
# 2:      IGKV2-28 -2.3961  FALSE
# 3:      ...      ...      ...

# order genes by lfc
genes <- genes[order(genes$fc, decreasing = TRUE),]
geneList <- genes$fc
names(geneList) <- genes$id

# load gene sets from database
gene_sets <- msigdbr(species = "Homo sapiens", category = "C5")
gene_sets  <- as.data.table(gene_sets)
pathways <- split(gene_sets$gene_symbol, gene_sets$gs_name)
# run fgsea with min, max size == 50, 500
fgsea_results <- fgsea(pathways = pathways, 
                       stats = geneList, 
                       minSize = 50, maxSize = 500)
\end{verbatim*}

\begin{figure}[H]
    \centering
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/go_mappingCompfgsea.png}
    \end{minipage}
    \hfill
    \begin{minipage}{0.49\textwidth}
        \centering
        \includegraphics[width=\textwidth]{./plots/ens_mappingCompgfgsea.png}
    \end{minipage}
    \caption{Comparing significantly enriched gene sets discovered by fgsea and our JAR, based on "-mappingtype".}
    \label{fig:go-mapping-fgsea}
\end{figure}

In Figure \ref{fig:go-mapping-fgsea}, it also seems as if the three \texttt{GOEntries} not discovered by 
\textit{fgsea} have likely been replaced or are simply non-existent in the \textit{msigdb C5}
gene set database. For both \textit{"-mappingtypes"}, our \textit{JAR} is able
to at least identify more than half of the gene sets discovered by \textit{fgsea}.

% ------------------------------------------------------------------------------





% ------------------------------------------------------------------------------
\printbibliography
% ------------------------------------------------------------------------------


\end{document}
